File: typescript/packages/extensions/src/sign-in-with-x/parse.ts
/**
 * Header parsing for SIWX extension
 *
 * Parses the SIGN-IN-WITH-X header from client requests.
 * Requires base64-encoded JSON per x402 v2 spec.
 */

import { Base64EncodedRegex, safeBase64Decode } from "@x402/core/utils";
import { SIWxPayloadSchema, type SIWxPayload } from "./types";

/**
 * Parse SIGN-IN-WITH-X header into structured payload.
 *
 * Expects base64-encoded JSON per x402 v2 spec (CHANGELOG-v2.md line 335).
 *
 * @param header - The SIGN-IN-WITH-X header value (base64-encoded JSON)
 * @returns Parsed SIWX payload
 * @throws Error if header is invalid or missing required fields
 *
 * @example
 * ```typescript
 * const header = request.headers.get('SIGN-IN-WITH-X');
 * if (header) {
 *   const payload = parseSIWxHeader(header);
 *   // payload.address, payload.signature, etc.
 * }
 * ```
 */
export function parseSIWxHeader(header: string): SIWxPayload {
  if (!Base64EncodedRegex.test(header)) {
    throw new Error("Invalid SIWX header: not valid base64");
  }

  const jsonStr = safeBase64Decode(header);

  let rawPayload: unknown;
  try {
    rawPayload = JSON.parse(jsonStr);
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new Error("Invalid SIWX header: not valid JSON");
    }
    throw error;
  }

  const parsed = SIWxPayloadSchema.safeParse(rawPayload);

  if (!parsed.success) {
    const issues = parsed.error.issues.map(i => `${i.path.join(".")}: ${i.message}`).join(", ");
    throw new Error(`Invalid SIWX header: ${issues}`);
  }

  return parsed.data;
}

File: python/x402/extensions/sign_in_with_x/parse.py
"""Header parsing for SIWX extension."""

from __future__ import annotations

import json
import re

from pydantic import ValidationError

from x402.http.utils import safe_base64_decode

from .types import SIWxPayload

_BASE64_RE = re.compile(r"^[A-Za-z0-9+/]*={0,2}$")


def parse_siwx_header(header: str) -> SIWxPayload:
    """Parse SIGN-IN-WITH-X header into structured payload."""
    if not _BASE64_RE.match(header):
        raise ValueError("Invalid SIWX header: not valid base64")

    json_str = safe_base64_decode(header)
    try:
        raw_payload = json.loads(json_str)
    except json.JSONDecodeError as e:
        raise ValueError("Invalid SIWX header: not valid JSON") from e

    try:
        return SIWxPayload.model_validate(raw_payload)
    except ValidationError as e:
        issues = ", ".join(
            f"{'.'.join(str(p) for p in err['loc'])}: {err['msg']}" for err in e.errors()
        )
        raise ValueError(f"Invalid SIWX header: {issues}") from e
